### PR TITLE
Resolution matrix

### LINKED ISSUE #6479: API: Deprecate PSF/CTF
@olafhauk recently @agramfort pointed me toward:

https://www.biorxiv.org/content/10.1101/672956v1

WDYT about deprecating MNE's PSF/CTF functionality, and instead change the existing examples to use your toolbox, with a link to your package?

If you can make a PR to update the examples to use your package, I can push commits to get the CIs (and requirements files, etc.) to pull in your package properly.

EDIT: Related: #4336

### PR DESCRIPTION
**Pull request by olafhauk**
closes #6479

### CODE REVIEW

File: mne/minimum_norm/resolution_matrix.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+    # apply inverse operator to identity matrix in order to get inverse matrix
+    # free orientation constraint not possible because apply_inverse would
+    # combine components
+
+    # pick_ori='normal' required because apply_inverse won't give separate
+    # orientations
+    # if ~inverse_operator['source_ori'] == FIFF.FIFFV_MNE_FIXED_ORI:
+    #     pick_ori = 'vector'
+    # else:
+    #     pick_ori = 'normal'
```
Comment (agramfort) [inline mne/minimum_norm/resolution_matrix.py]: cleanup?

File: mne/minimum_norm/tests/test_resolution_matrix.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    # fixed source orientation
+    inverse_operator_fxd = mne.minimum_norm.make_inverse_operator(
+        info=evoked.info, forward=forward, noise_cov=noise_cov, loose=0.,
+        depth=None, fixed=True)
+
+    # regularisation parameter based on SNR
+    snr = 3.0
+    lambda2 = 1.0 / snr ** 2
+
+    print('###\nComputing resolution matrices.\n###')
```
Comment (agramfort) [inline mne/minimum_norm/tests/test_resolution_matrix.py]: remove

File: mne/minimum_norm/tests/test_resolution_matrix.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    # Does sLORETA have zero DLE for columns?
+    assert_equal(maxidxs, goodidxs)
+
+    # MNE resolution matrices symmetric?
+    assert_array_almost_equal(rm_mne, rm_mne.T)
+    assert_array_almost_equal(rm_mne_free, rm_mne_free.T)
+
+    return rm_mne_free, rm_mne, rm_lor
+
+# rm_mne_free, rm_mne, rm_lor = test_resolution_matrix()
```
Comment (agramfort) [inline mne/minimum_norm/tests/test_resolution_matrix.py]: cleanup

File: mne/minimum_norm/resolution_matrix.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+
+from mne import pick_channels
+from mne.utils import logger
+from mne.forward.forward import is_fixed_orient, convert_forward_solution
+
+from mne.evoked import EvokedArray
+from mne.minimum_norm import apply_inverse
+
+
+def make_resolution_matrix(fwd, invop, method, lambda2):
```
Comment (jasmainak) [inline mne/minimum_norm/resolution_matrix.py]: @johnsam7 you might want to take a look at this

**Comment by larsoner** [comment]
@olafhauk let me know when you've addressed existing comments and I'm happy to take a look

File: mne/minimum_norm/resolution_matrix.py [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+
+from mne import pick_channels, EvokedArray
+from mne.utils import logger
+from mne.forward.forward import is_fixed_orient, convert_forward_solution
+
+from mne.minimum_norm import apply_inverse
+from mne.beamformer import make_lcmv
+
+
+def make_resolution_matrix(forward, inverse_operator, method, lambda2):
```
Comment (agramfort) [inline mne/minimum_norm/resolution_matrix.py]: use the same defaults as apply_inverse so you can just do

`make_resolution_matrix(forward, inverse_operator)`

File: mne/minimum_norm/resolution_matrix.py [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+    inverse_operator: Instance of InverseOperator
+        Inverse operator.
+    method: str
+        Inverse method to use (MNE, dSPM, sLORETA).
+    lambda2: float
+        The regularisation parameter.
+
+    Returns
+    -------
+        resmat: numpy array, shape (n_dipoles, n_dipoles)
```
Comment (agramfort) [inline mne/minimum_norm/resolution_matrix.py]:
```suggestion
    resmat: numpy array, shape (n_dipoles, n_dipoles)
```

File: mne/minimum_norm/resolution_matrix.py [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+
+    dims = invmat.shape
+
+    # remove columns for bad channels
+    # take into account it may be 3D array
+    invmat = np.delete(invmat, ch_idx_bads, axis=len(dims) - 1)
+
+    # if 3D array, i.e. multiple values per location (fixed and loose),
+    # reshape into 2D array
+    if len(dims) == 3:
```
Comment (agramfort) [inline mne/minimum_norm/resolution_matrix.py]:
```suggestion
    if invmat.ndim == 3:
```

File: mne/minimum_norm/resolution_matrix.py [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+                              method=method, pick_ori=pick_ori)
+
+    # turn source estimate into numpty array
+    invmat = invmat_op.data
+
+    dims = invmat.shape
+
+    # remove columns for bad channels
+    # take into account it may be 3D array
+    invmat = np.delete(invmat, ch_idx_bads, axis=len(dims) - 1)
```
Comment (agramfort) [inline mne/minimum_norm/resolution_matrix.py]:
```suggestion
    invmat = np.delete(invmat, ch_idx_bads, axis=invmat.ndim - 1)
```

File: mne/minimum_norm/resolution_matrix.py [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+
+    # if 3D array, i.e. multiple values per location (fixed and loose),
+    # reshape into 2D array
+    if len(dims) == 3:
+        v0o1 = invmat[0, 1].copy()
+        v3o2 = invmat[3, 2].copy()
+        invmat = invmat.reshape(dims[0] * dims[1], dims[2])
```
Comment (agramfort) [inline mne/minimum_norm/resolution_matrix.py]:
```suggestion
        shape = invmat.shape
        invmat = invmat.reshape(shape[0] * shape[1], shape[2])
```

then remove dims. I don't like the word `dims` as shape is the common name.

File: mne/minimum_norm/tests/test_resolution_matrix.py [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+    rm_lor_abs = np.abs(rm_lor)
+
+    # get maxima per column
+    maxidxs = rm_lor_abs.argmax(axis=0)
+
+    # create array with the expected stepwise increase in maximum indices
+    goodidxs = np.arange(0, len(maxidxs), 1)
+
+    # Tests
+    # Does sLORETA have zero DLE for columns?
```
Comment (agramfort) [inline mne/minimum_norm/tests/test_resolution_matrix.py]:
```suggestion
    # Does sLORETA have zero localization error for columns?
```

avoid acronyms as they are harder to understand for new comers

File: mne/minimum_norm/tests/test_resolution_matrix.py [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+    assert_equal(maxidxs, goodidxs)
```
Comment (agramfort) [inline mne/minimum_norm/tests/test_resolution_matrix.py]:
```suggestion
    assert_array_equal(maxidxs, goodidxs)
```

File: mne/minimum_norm/tests/test_resolution_matrix.py [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+
+    # MNE resolution matrices symmetric?
+    assert_array_almost_equal(rm_mne, rm_mne.T)
+    assert_array_almost_equal(rm_mne_free, rm_mne_free.T)
+
+#    return rm_mne_free, rm_mne, rm_lor, resmat_lcmv, filters
+
+# rm_mne_free, rm_mne, rm_lor, resmat_lcmv, filters = test_resolution_matrix()
```
Comment (agramfort) [inline mne/minimum_norm/tests/test_resolution_matrix.py]: remove

File: mne/minimum_norm/resolution_matrix.py [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+
+    logger.info('Dimensions of resolution matrix: %d by %d.' % resmat.shape)
+
+    return resmat
+
+
+def _convert_forward_match_inv(fwd, inv):
+    """Helper to ensure forward and inverse operators match."""
+    # did inverse operator use fixed orientation?
+    is_fixed_inv = inv['eigen_leads']['data'].shape[0] == inv['nsource']
```
Comment (agramfort) [inline mne/minimum_norm/resolution_matrix.py]: same here use:

inverse_operator['source_ori'] != FIFF.FIFFV_MNE_FREE_ORI

or better make a private function you can use at the 2 places

File: mne/minimum_norm/resolution_matrix.py [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+    if is_fixed_inv:
+        pick_ori = None
+    else:
+        pick_ori = 'vector'
+
+    # columns for bad channels will be zero
+    invmat_op = apply_inverse(ev_id, inverse_operator, lambda2=lambda2,
+                              method=method, pick_ori=pick_ori)
+
+    # turn source estimate into numpty array
```
Comment (agramfort) [inline mne/minimum_norm/resolution_matrix.py]:
```suggestion
    # turn source estimate into numpy array
```

File: mne/minimum_norm/tests/test_resolution_matrix.py [hunk 15]
```diff
@@ -0,0 +1,107 @@
+"""
+Test the following properties of the resolution matrix.
+
+Resolution matrix is symmetrical for MNE (fixed and free source orientation).
+Resolution matrix has zero dipole localisation error for columns (PSFs).
+"""
+
```
Comment (agramfort) [inline mne/minimum_norm/tests/test_resolution_matrix.py]: add license and author like here https://github.com/mne-tools/mne-python/blob/master/mne/tests/test_event.py#L2

File: mne/minimum_norm/tests/test_resolution_matrix.py [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+fname_fwd = op.join(data_path, 'MEG', 'sample',
+                    'sample_audvis_trunc-meg-eeg-oct-4-fwd.fif')
+fname_cov = op.join(data_path, 'MEG', 'sample', 'sample_audvis_trunc-cov.fif')
+
+# get functions for resolution matrix, metrics etc.
+
+
+@testing.requires_testing_data
+def test_resolution_matrix():
+    """Test whether MNE's resolution matrix is symmetric."""
```
Comment (agramfort) [inline mne/minimum_norm/tests/test_resolution_matrix.py]: you test more than this. Just say
```suggestion
    """Test make_resolution_matrix function."""
```

File: mne/minimum_norm/resolution_matrix.py [hunk 17]
```diff
@@ -0,0 +1,225 @@
...
+        ----------
+        resmat: array, shape (n_dipoles, n_dipoles)
+            Forward Operator.
+        src: Source Space
+            Source space used to compute resolution matrix.
+        idx: list of int
+            Vertex indices for which PSFs or CTFs to produce.
+        func: str ('psf' | 'ctf')
+            Whether to produce PSFs or CTFs.
+        norm: Bool
```
Comment (agramfort) [inline mne/minimum_norm/resolution_matrix.py]:
```suggestion
        norm: bool
```

File: mne/minimum_norm/resolution_matrix.py [hunk 18]
```diff
@@ -0,0 +1,225 @@
...
+            Whether to normalise to maximum across all PSFs and CTFs (default:
+            False)
+
+    Returns
+        -------
+        stc: STC object
```
Comment (agramfort) [inline mne/minimum_norm/resolution_matrix.py]:
```suggestion
        stc: instance of SourceEstimate
```

**Comment by agramfort** [comment]
looks good https://15975-1301584-gh.circle-artifacts.com/0/dev/auto_examples/inverse/plot_psf_ctf_vertices.html

**Comment by agramfort** [comment]
thx @olafhauk

**Pull request merged**